nym-network-monitor/src/http.rs

```diff
@@ -0,0 +1,106 @@
+
+pub struct HttpServer {
+    listener: SocketAddr,
+    cancel: CancellationToken,
+}
+
+#[derive(OpenApi)]
+#[openapi(
+    paths(
+        crate::handlers::accounting_handler,
+        crate::handlers::graph_handler,
+        crate::handlers::mermaid_handler,
+        crate::handlers::mix_dot_handler,
+        crate::handlers::node_stats_handler,
+        crate::handlers::recv_handler,
+        crate::handlers::send_handler,
+        crate::handlers::sent_handler,
+        crate::handlers::all_nodes_stats_handler,
+    ),
+    components(schemas(
+        FragmentHeader,
+        FragmentsReceived,
+        FragmentsSent,
+        NetworkAccount,
+        NetworkAccountStats,
+        NodeStats,
+        ReceivedFragment,
+        SentFragment,
+    ))
+)]
+struct ApiDoc;
+
+#[derive(Clone)]
+pub struct AppState {
+    clients: ClientsWrapper,
+}
+
+impl AppState {
+    pub fn clients(&self) -> &ClientsWrapper {
+        &self.clients
+    }
+}
+
+impl HttpServer {
+    pub fn new(listener: SocketAddr, cancel: CancellationToken) -> Self {
+        HttpServer { listener, cancel }
+    }
+
+    pub async fn run(self, clients: ClientsWrapper) -> anyhow::Result<()> {
+        let n_clients = clients.read().await.len();
+        let state = AppState { clients };
+        let app = Router::new()
+            .route("/v1/send", post(send_handler).with_state(state))
+            .merge(SwaggerUi::new("/v1/ui").url("/v1/docs/openapi.json", ApiDoc::openapi()))
+            .route("/v1/accounting", get(accounting_handler))
+            .route("/v1/sent", get(sent_handler))
+            .route("/v1/dot/:mix_id", get(mix_dot_handler))
+            .route("/v1/dot", get(graph_handler))
+            .route("/v1/mermaid", get(mermaid_handler))
+            .route("/v1/stats", get(stats_handler))
+            .route("/v1/node_stats/:mix_id", get(node_stats_handler))
+            .route("/v1/node_stats", get(all_nodes_stats_handler))
+            .route("/v1/received", get(recv_handler));
+        let listener = tokio::net::TcpListener::bind(self.listener).await?;
+
+        let shutdown_future = self.cancel.cancelled();
+        let server_future = axum::serve(listener, app).into_future();
+
+        info!("##########################################################################################");
+        info!("######################### HTTP server running, with {} clients ############################################", n_clients);
+        info!("##########################################################################################");
+
+        tokio::select! {
+            _ = shutdown_future => {
+                info!("received shutdown");
+            }
+            res = server_future => {
+                info!("the http server has terminated");
+                if let Err(err) = res {
+                    info!("with the following error: {err}");
+                    return Err(err.into())
+                }
+            }
+        }
```

Comment:
This looks like a forced shutdown of axum server: when `shutdown_future` branch of `select` completes, other branch (`server_future`) is cancelled. 

Besides, there is a more ergonomic way to use a cancellation token future with Axum server, where you don't need to manually "cancel" server branch after you await the cancellation token:

```rust
axum::serve(listener, app).with_graceful_shutdown(self.cancel.cancelled())
```


`.with_graceful_shutdown` accepts a future which, when resolved, starts graceful shutdown. 

`CancellationToken::cancelled()` offers just that kind of future so they work hand in hand. 

This way, you just issue a cancel signal on the token, and `tokio` will signal axum server to initiate a (graceful) shutdown.

Reply:
Nice one!